tempesta_fw/http.c

```diff
@@ -2811,6 +2815,58 @@ tfw_http_chop_skb(TfwHttpMsg *hm, struct sk_buff *skb,
 	return ss_skb_chop_head_tail(hm->msg.skb_head, skb, off, trail);
 }
 
+static TfwStr
+tfw_http_get_ip_from_xff(TfwHttpReq *req)
+{
+	TfwStr s_xff, s_ip, *c, *end;
+	unsigned int nchunks;
+
+	/*
+	 * If a client work through a forward proxy, then a proxy can pass it's
+	 * IP address by the first value in X-Forwarded-For
+	 */
+	s_xff = req->h_tbl->tbl[TFW_HTTP_HDR_X_FORWARDED_FOR];
+	s_ip = tfw_str_next_str_val(&s_xff);
+	nchunks = 0;
+	TFW_STR_FOR_EACH_CHUNK(c, &s_ip, end) {
+		if (!(c->flags & TFW_STR_VALUE))
+			break;
+		nchunks++;
+	}
+	s_ip.nchunks = nchunks;
+
+	return s_ip;
+}
+
+static int
+tfw_http_req_client_link(TfwConn *conn, TfwHttpReq *req)
+{
+	TfwStr s_ip, s_user_agent, *ua;
+	TfwAddr addr;
+	TfwClient *cli, *conn_cli;
+
+	s_ip = tfw_http_get_ip_from_xff(req);
+	if (!TFW_STR_EMPTY(&s_ip)) {
+		if (tfw_addr_pton(&s_ip, &addr) != 0)
+			return TFW_BLOCK;
+
+		conn_cli = (TfwClient *)conn->peer;
+		ua = &req->h_tbl->tbl[TFW_HTTP_HDR_USER_AGENT];
+		tfw_http_msg_clnthdr_val(ua, TFW_HTTP_HDR_USER_AGENT,
+					 &s_user_agent);
+		cli = tfw_client_obtain(conn_cli->addr, &addr, &s_user_agent,
+					NULL);
+		if (cli) {
+			if (cli != conn_cli)
+				req->peer = cli;
```

Comment:
`req->peer` most probably was obtained by calling `tfw_client_obtain()`. Since this assignment loses `req->peer`, shouldn't it release it by calling `tfw_client_put()`?

Reply:
Confused with `conn->peer` which is set for every connection. Unlike that, `req->peer` is only set here, so there is no need to release previous value.